File: problems/Valid Perfect Square/isPerfectSquare.cs
public class Solution {
    public bool IsPerfectSquare(int num) {
        if (2 > num) {
            return true;
        }

        long x = num >> 1;

        while (num < x * x) {
            x = (x + num / x) >> 1;
        }

        return (num == x * x);
    }
}

File: problems/Valid Perfect Square/isPerfectSquare.java
class Solution {
    public boolean isPerfectSquare(int num) {
        if (num < 2) {
            return true;
        }

        long left = 0;
        long right = num;

        while (left + 1 < right) {
            long mid = left + (right - left) / 2;

            if (mid * mid == num) {
                return true;
            } else if (mid * mid < num) {
                left = mid;
            } else {
                right = mid;
            }
        }

        return false;
    }
}
